src/debug/dart_debug_impl.ts

```diff
@@ -738,6 +744,65 @@ export class DartDebugSession extends DebugSession {
 		}
 	}
 
+	private async updateCoverage(scriptUris: string[]): Promise<void> {
+		const coverageReport = await this.getCoverageReport(scriptUris);
+
+		// Unwrap tokenPos into real locations.
+		const coverageData = coverageReport.map((r) => ({
+			hits: r.hits.map((h) => this.resolveFileLocation(r.script, h)),
+			scriptUri: r.script.uri,
+		}));
+
+		this.sendEvent(new Event("dart.coverage", coverageData));
+	}
+
+	private async getCoverageReport(scriptUris: string[]): Promise<Array<{ script: VMScript, tokenPosTable: number[][], hits: number[] }>> {
+		// TODO: Do we need to do all of these requests every time? Can we stack the loaded scripts?
+		const result = await this.observatory.getVM();
+		const vm = result.result as VM;
+
+		const promises: Array<Promise<DebuggerResult>> = [];
+
+		const isolatePromises = vm.isolates.map((isolateRef) => this.observatory.getIsolate(isolateRef.id));
+		const isolatesResponses = await Promise.all(isolatePromises);
+		const isolates = isolatesResponses.map((response) => response.result as VMIsolate);
+
+		const results: Array<{ script: VMScript, tokenPosTable: number[][], hits: number[] }> = [];
+		const scriptPaths = scriptUris.map((s) => uriToFilePath(s));
+		for (const isolate of isolates) {
+			const libraryPromises = isolate.libraries.map((library) => this.observatory.getObject(isolate.id, library.id));
+			const libraryResponses = await Promise.all(libraryPromises);
+			const libraries = libraryResponses.map((response) => response.result as VMLibrary);
+
+			const scriptRefs = flatMap(libraries, (library) => library.scripts);
+
+			// Filter scripts to the ones we care about.
+			// TODO: Do we need to getPossibleSourceUris() here?
+			const scripts = scriptRefs.filter((s) => scriptPaths.indexOf(uriToFilePath(s.uri)) !== -1);
+
+			for (const scriptRef of scripts) {
+				const script = (await this.observatory.getObject(isolate.id, scriptRef.id)).result as VMScript;
+				try {
+					const report = await this.observatory.getSourceReport(isolate, [SourceReportKind.Coverage], scriptRef);
```

Comment:
@devoncarew Currently I'm sending a LOT of requests to the VM for this:

- getVM
- for each isolate, getIsolate
- for each library in each isolate, getLibrary
- for each script in each library in each isolate, getScript
- for each script in each library in each isolate, getSourceReport

I'm not sure how safe it would be to cache any of these (and we don't really have any caching in the debug adapter right now, except for the storeData/getData lookup for passing back "references" to Code), since presumably new isolates can be created, new libraries can be loaded, etc.

Looking at your code, I think you're doing the same (`getScriptRefFor` seems to wrap some of it up). I'm not sure if that's just because it's work-in-progress or it's the correct thing to do though.

Reply:
So:

- the VM calls are fast :) (modulo the getSourceReport() call)
- I re-request everything each time I query for coverage, but cache during that call (for things like script token positions)
- I only query for the current active file, given the cost of getSourceReport

Reply:
Ok, sounds similar :) I'm not explicitly caching anything, but if there's only one isolate and a script can only exist in one library, then I don't think there will be duplicated requests for anything? (I do reuse the same `Script` instance for mapping all tokens in the response since it's in scope with the results).

I'm currently doing "all visible files" so that if a user has a split view, it'll update each (rather than them appearing/disappearing as they move focus between the two panes, which would also be increased calls) but once I've got everything working I can check timings and see how things are.

Thanks!

Reply:
Makes sense re: caching - my impl was written before we changed to a very targeted view of the markers (I think you're correct in that the 'cache' would only be hit once).

> I'm currently doing "all visible files" so that if a user has a split view

This will likely be performant - requesting for a small number of files will be fast. The general call - getSourcereport w/o qualifying the script - requests for all 600-700; that is slow.